src/coreclr/jit/optimizer.cpp

```diff
@@ -4951,17 +4439,315 @@ void Compiler::optFindNewLoops()
 // Remarks:
 //   Guarantees that all natural loops have preheaders.
 //
-bool Compiler::optCanonicalizeLoops(FlowGraphNaturalLoops* loops)
+bool Compiler::optCanonicalizeLoops()
 {
     bool changed = false;
-    for (FlowGraphNaturalLoop* loop : loops->InReversePostOrder())
+    for (FlowGraphNaturalLoop* loop : m_loops->InReversePostOrder())
     {
         changed |= optCreatePreheader(loop);
     }
 
     return changed;
 }
 
+//-----------------------------------------------------------------------------
+// optCompactLoops: Compact loops to make their loop blocks lexical if possible.
+//
+// Returns:
+//   True if the flow graph was changed.
+//
+bool Compiler::optCompactLoops()
+{
+    bool changed = false;
+    for (FlowGraphNaturalLoop* loop : m_loops->InReversePostOrder())
+    {
+        if (!loop->GetHeader()->HasFlag(BBF_OLD_LOOP_HEADER_QUIRK))
+        {
+            continue;
+        }
+
+        changed |= optCompactLoop(loop);
+    }
+
+    return changed;
+}
+
+//-----------------------------------------------------------------------------
+// optCompactLoop: Compact a specific loop.
+//
+// Parameters:
+//   loop - The loop
+//
+// Returns:
+//   True if the flow graph was changed.
+//
+bool Compiler::optCompactLoop(FlowGraphNaturalLoop* loop)
+{
+    BasicBlock* insertionPoint = nullptr;
+
+    BasicBlock* top           = loop->GetLexicallyTopMostBlock();
+    unsigned    numLoopBlocks = loop->NumLoopBlocks();
+
+    BasicBlock* cur              = top;
+    bool        changedFlowGraph = false;
+    while (numLoopBlocks > 0)
+    {
+        if (loop->ContainsBlock(cur))
+        {
+            numLoopBlocks--;
+            cur = cur->Next();
+            continue;
+        }
+
+        BasicBlock* lastNonLoopBlock = cur;
+        while (true)
+        {
+            // Should always have a "bottom" block of the loop where we stop.
+            assert(lastNonLoopBlock->Next() != nullptr);
+            if (loop->ContainsBlock(lastNonLoopBlock->Next()))
+            {
+                break;
+            }
+
+            lastNonLoopBlock = lastNonLoopBlock->Next();
+        }
+
+        if (insertionPoint == nullptr)
+        {
+            insertionPoint = optFindLoopCompactionInsertionPoint(loop, top);
+        }
+
+        BasicBlock* previous      = cur->Prev();
+        BasicBlock* nextLoopBlock = lastNonLoopBlock->Next();
+        assert(previous != nullptr);
+        if (!BasicBlock::sameEHRegion(previous, nextLoopBlock) || !BasicBlock::sameEHRegion(previous, insertionPoint))
```

Comment:
fyi: this code is basically pre-existing, but there's an issue for fixing/improving it: https://github.com/dotnet/runtime/issues/85550